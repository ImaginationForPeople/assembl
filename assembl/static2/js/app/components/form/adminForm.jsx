// @flow
/*
  Form with save button and a popup that notifies user if he leaves the form with unsaved data
*/
import * as React from 'react';
import { I18n } from 'react-redux-i18n';
import { type Route, type Router, withRouter } from 'react-router';

import SaveButton from '../administration/saveButton';

type Props = {
  children: React.Node,
  handleSubmit: Function,
  pristine: boolean,
  routes: Array<Route>,
  router: Router,
  submitting: boolean
};

export class DumbAdminForm extends React.Component<Props> {
  componentDidMount() {
    const route = this.props.routes.slice(-1)[0];
    this.props.router.setRouteLeaveHook(route, this.routerWillLeave);
  }

  componentWillUnmount() {
    const route = this.props.routes.slice(-1)[0];
    this.props.router.setRouteLeaveHook(route, null);
  }

  routerWillLeave = () => {
    const { pristine, submitting } = this.props;
    if (!pristine && !submitting) {
      return I18n.t('administration.confirmUnsavedChanges');
    }

    return null;
  };

  render() {
    const { children, handleSubmit, pristine, submitting } = this.props;
    return (
      <form onSubmit={handleSubmit}>
        <SaveButton disabled={pristine || submitting} saveAction={handleSubmit} />
        {children}
      </form>
    );
  }
}

export default withRouter(DumbAdminForm);