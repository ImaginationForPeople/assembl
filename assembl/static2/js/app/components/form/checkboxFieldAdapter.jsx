// @flow
import React from 'react';
import { type FieldRenderProps } from 'react-final-form';
import { Checkbox, FormGroup } from 'react-bootstrap';

import Error from './error';
import { getValidationState } from './utils';

type Props = {
  label: string
} & FieldRenderProps;

const CheckboxFieldAdapter = ({ input, label, meta: { error, touched } }: Props) => (
  <FormGroup controlId={input.name} validationState={getValidationState(error, touched)}>
    <Checkbox {...input}>{label}</Checkbox>
    <Error name={input.name} />
  </FormGroup>
);

export default CheckboxFieldAdapter;