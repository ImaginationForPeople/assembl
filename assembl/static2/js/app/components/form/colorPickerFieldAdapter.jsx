// @flow
import * as React from 'react';
import { BlockPicker as ColorPicker } from 'react-color';
import { FormGroup } from 'react-bootstrap';

import { type FieldRenderProps } from 'react-final-form';
import { getValidationState } from './utils';
import { pickerColors } from '../../constants';
import Error from './error';

type Props = {
  name: string,
  label: string,
  input: {
    name: string,
    onChange: Function,
    value: string
  }
} & FieldRenderProps;

const ColorPickerFieldAdapter = ({ meta: { error, touched }, input: { name, onChange, value }, label }: Props) => {
  const onChangeColor = (v) => {
    onChange(v.hex);
  };
  return (
    <FormGroup controlId={name} validationState={getValidationState(error, touched)}>
      <label htmlFor={`color-picker-${name}`}>{label}</label>
      <ColorPicker colors={pickerColors} color={value} width="100%" id={`color-picker-${name}`} triangle="hide" className="no-box-shadow" onChange={onChangeColor} />
      <Error name={name} />
    </FormGroup>
  );
};

export default ColorPickerFieldAdapter;