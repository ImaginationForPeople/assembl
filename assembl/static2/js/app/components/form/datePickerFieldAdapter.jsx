// @flow

/*
  DatePicker adapter for react-final-form that supports the following form:
  { time: moment().utc() }

  Errors are managed by a Mutator on the Form itself. As a result, to show errors
  on DatePickerFieldAdapter, must add the `setFieldTouched` from final-form-set-field-touched
  along with passing the form <FormApi> prop
*/
import DatePicker from 'react-datepicker';
import 'react-datepicker/dist/react-datepicker.css';
import * as React from 'react';
import { ControlLabel, FormGroup } from 'react-bootstrap';
import { type FieldRenderProps } from 'react-final-form';
import { type FormApi } from 'final-form';
import type { DatePickerType, DatePickerValue } from './types.flow';
import Error from './error';
import { getValidationState } from './utils';

type Props = {
  editLocale: string,
  picker?: DatePickerType,
  placeHolder: string,
  showTime: boolean,
  hasConflictingDates: boolean,
  required: boolean,
  input: {
    name: string,
    onChange: ({ time: moment$Moment }) => void,
    value: DatePickerValue
  },
  onDateChange: ?(moment$Moment) => void,
  form: FormApi,
  children?: React.Node,
  dateFormat?: string
} & FieldRenderProps;

const DatePickerFieldAdapter = ({
  editLocale,
  picker,
  placeHolder,
  showTime,
  input: { name, value, onChange },
  meta: { error, touched },
  required,
  hasConflictingDates,
  onDateChange,
  children,
  form,
  dateFormat,
  ...rest
}: Props) => {
  const decoratedPlaceholder = required ? `${placeHolder} *` : placeHolder;
  const onLocalizedChange = (e: moment$Moment): void => {
    if (e !== value.time) {
      if (form) {
        form.mutators.setFieldTouched(name, true);
      }
      if (onDateChange) {
        onDateChange(e);
      }
    }
    return onChange({ time: e });
  };

  return (
    <FormGroup validationState={getValidationState(error, touched)}>
      <div className="date-picker-field">
        <ControlLabel htmlFor={`date-picker-${name}`} className="datepicker-label">
          {picker &&
            picker.pickerType && <div className={`date-picker-type ${picker.pickerClasses || ''}`}>{picker.pickerType}</div>}
          <DatePicker
            placeholderText={decoratedPlaceholder}
            selected={value.time}
            id={`date-picker-${name}`}
            onChange={onLocalizedChange}
            showTimeSelect={showTime}
            dateFormat={dateFormat}
            locale={editLocale}
            shouldCloseOnSelect
            className={hasConflictingDates ? 'warning' : ''}
            {...rest}
          />
          <div className="icon-schedule-container">
            <span className="assembl-icon-schedule grey" />
          </div>
        </ControlLabel>
        <Error name={name} />
        {children || null}
      </div>
    </FormGroup>
  );
};

DatePickerFieldAdapter.defaultProps = {
  showTime: false,
  dateFormat: 'LLL'
};

export default DatePickerFieldAdapter;