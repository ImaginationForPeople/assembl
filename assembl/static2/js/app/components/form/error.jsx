// @flow
import React from 'react';
import { HelpBlock } from 'react-bootstrap';
import { Field } from 'react-final-form';

type Props = {
  name: string
};

const Error = ({ name }: Props) => (
  <Field
    name={name}
    subscription={{ touched: true, error: true }}
    render={({ meta: { touched, error } }) => (touched && error ? <HelpBlock>{error}</HelpBlock> : null)}
  />
);

export default Error;