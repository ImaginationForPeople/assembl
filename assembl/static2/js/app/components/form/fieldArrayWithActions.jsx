// @flow
import * as React from 'react';
import { Button, OverlayTrigger } from 'react-bootstrap';
import { Translate } from 'react-redux-i18n';
import { FieldArray, type FieldArrayRenderProps } from 'react-final-form-arrays';
import classNames from 'classnames';
import range from 'lodash/range';
import isEmpty from 'lodash/isEmpty';

import { displayModal, closeModal } from '../../utils/utilityManager';
import { upTooltip, downTooltip } from '../common/tooltips';
import { createRandomId, getDomElementOffset } from '../../utils/globalFunctions';
import { MAX_TREE_FORM_LEVEL } from '../../constants';

type ConfirmationMessageType = {
  field: Object,
  index: string
};

type Props = {
  usePanels: boolean,
  renderFields: Function,
  titleMsgId?: string,
  renderTitleMsg: ({ titleMsgId: string, idx: number, fieldValue: mixed }) => React.Node,
  tooltips: {
    addTooltip: (props: { level: number }) => React.Node,
    deleteTooltip: () => React.Node,
    deleteDisabled?: () => React.Node
  },
  withSeparators: boolean,
  subFieldName?: string,
  isTree: boolean,
  level: number,
  maxLevel: number,
  minItems: number,
  parents: Array<number>,
  parentId: string,
  confirmDeletion: boolean,
  confirmDeletionMessages: {
    confirmDeletionTitle: (props: ConfirmationMessageType) => React.Node,
    confirmDeletionBody: (props: ConfirmationMessageType) => React.Node
  },
  errors?: Object
};

type State = {
  activePanel: number | null
};

type FieldsProps = {
  fields: FieldArrayRenderProps,
  onAdd?: (id: string, parentId: string, index: number) => void,
  onRemove?: (id: string) => void,
  onUp?: (id: string, parentId: string, index: number, targetIndex: number) => void,
  onDown?: (id: string, parentId: string, index: number, targetIndex: number) => void
} & Props;

type FieldArrayProps = {
  name: string
} & Props;

export function confirmDeletionModal(title: React.Node, body: React.Node, remove: () => void) {
  const footer = [
    <Button key="cancel" onClick={closeModal} className="button-cancel button-dark">
      <Translate value="cancel" />
    </Button>,
    <Button
      key="delete"
      onClick={() => {
        remove();
        closeModal();
      }}
      className="button-submit button-dark"
    >
      <Translate value="delete" />
    </Button>
  ];
  return displayModal(title, body, true, footer);
}

export class Fields extends React.PureComponent<FieldsProps, State> {
  constructor(props: FieldsProps) {
    super(props);
    this.initialize();
  }

  state = { activePanel: null };

  initialize = () => {
    const { fields, minItems, level, onAdd, parentId } = this.props;
    const fieldsLength = fields.value ? fields.value.length : 0;
    if (level === 0 && fieldsLength < minItems) {
      range(0, minItems - fieldsLength).forEach(() => {
        const id = createRandomId();
        if (onAdd) {
          const index = (fields.length || 0) + 1;
          setTimeout(() => onAdd(id, parentId, index), 2000);
        }
        fields.push({ id: id });
      });
    }
  };

  down = (index: number) => {
    const { fields, onDown, parentId } = this.props;
    if (onDown) {
      onDown(fields.value[index].id, parentId, index, index + 1);
    }
    fields.swap(index, index + 1);
  };

  up = (index: number) => {
    const { fields, onUp, parentId } = this.props;
    if (onUp) {
      onUp(fields.value[index].id, parentId, index, index - 1);
    }
    fields.swap(index, index - 1);
  };

  removedId: ?number = null;

  remove = (index: number) => {
    const { fields, onRemove, isTree, subFieldName } = this.props;
    const fieldValue = fields.value[index];
    const children = subFieldName && isTree && fieldValue[subFieldName];
    if (!children || children.length === 0) {
      this.removedId = fieldValue.id;
      if (onRemove) {
        onRemove(fieldValue.id);
      }
      fields.remove(index);
    }
  };

  add = () => {
    const { fields, onAdd, parentId, usePanels } = this.props;
    const id = createRandomId();
    const idx = fields.length || 0;
    if (onAdd) {
      onAdd(id, parentId, idx);
    }
    fields.push({ id: id });
    if (usePanels) {
      this.setActivePanel(idx);
    }
  };

  setActivePanel = (idx: number) => {
    this.setState(
      prevState => ({ activePanel: prevState.activePanel === idx ? null : idx }),
      () => {
        setTimeout(() => {
          const panel = document.getElementById(`panel${idx}`);
          const scrollY = panel ? getDomElementOffset(panel).top : 0;
          window.scrollTo({ top: scrollY - 170, left: 0, behavior: 'smooth' });
        }, 20);
      }
    );
  };

  render() {
    const {
      fields,
      renderFields,
      titleMsgId,
      renderTitleMsg,
      isTree,
      level,
      subFieldName,
      maxLevel,
      minItems,
      withSeparators,
      confirmDeletion,
      parents,
      tooltips: { addTooltip, deleteTooltip, deleteDisabled },
      confirmDeletionMessages: { confirmDeletionTitle, confirmDeletionBody },
      onAdd,
      onRemove,
      onUp,
      onDown,
      usePanels,
      errors
    } = this.props;
    /* Hack to fix issue with richtext field:
    when clicking on delete action, the value of richtext field becomes '' and so
    new EditorState.createEmpty() is created triggering onChange event on first render of the field)
    preventing the whole block to be removed */
    const removedId = this.removedId;
    const isRoot = level === 0;
    const className = level > 0 ? 'form-branch' : 'form-tree';
    const addBtnTop = isTree && !isRoot;
    const displayAddBtn = !isTree || (isTree && level < maxLevel);
    const addBtn = displayAddBtn ? (
      <OverlayTrigger placement="top" overlay={addTooltip({ level: level + 1 })}>
        <div onClick={this.add} className={classNames('plus margin-l', { 'form-tree-item': isTree })}>
          +
        </div>
      </OverlayTrigger>
    ) : null;
    return (
      <div className={classNames({ [className]: isTree })}>
        {addBtnTop ? addBtn : null}
        {fields.map((fieldname, idx) => {
          if (fields.value[idx].id === removedId) {
            return null;
          }
          const fieldValue = fields.value[idx];
          const hasChildren = fieldValue.children && fieldValue.children.length;
          const enableDeleteBtn = fields.length > minItems && ((isTree && !hasChildren) || !isTree);
          const hideDeleteBtn = isRoot && !enableDeleteBtn;
          const displaySeparator = withSeparators && (!isTree || (isRoot && idx === fields.length - 1));
          const indexes = [...parents, idx + 1];
          const fieldIndex = indexes.join('.');
          const removeField = () => this.remove(idx);
          const removeConfirmationData = { field: fieldValue, index: fieldIndex };
          const removeWithConfirmation = confirmDeletion
            ? () =>
              confirmDeletionModal(
                confirmDeletionTitle(removeConfirmationData),
                confirmDeletionBody(removeConfirmationData),
                removeField
              )
            : removeField;
          return (
            <div
              className={classNames('form-container', {
                'panel panel-default': usePanels,
                'panel-error': this.state.activePanel !== idx && errors && !isEmpty(errors[idx])
              })}
              id={usePanels ? `panel${idx}` : undefined}
              key={fieldname}
            >
              {titleMsgId ? (
                <div
                  className={classNames({
                    title: !usePanels,
                    'panel-heading pointer': usePanels,
                    left: true
                  })}
                  onClick={usePanels ? () => this.setActivePanel(idx) : undefined}
                >
                  {renderTitleMsg({ titleMsgId: titleMsgId, idx: idx, fieldValue: fieldValue })}
                </div>
              ) : null}
              <div className="pointer right">
                <div className="inline">
                  {idx < fields.length - 1 ? (
                    <OverlayTrigger placement="top" overlay={downTooltip}>
                      <Button onClick={() => this.down(idx)} className="admin-icons">
                        <span className="assembl-icon-down-bold grey" />
                      </Button>
                    </OverlayTrigger>
                  ) : null}
                  {idx > 0 ? (
                    <OverlayTrigger placement="top" overlay={upTooltip}>
                      <Button onClick={() => this.up(idx)} className="admin-icons">
                        <span className="assembl-icon-up-bold grey" />
                      </Button>
                    </OverlayTrigger>
                  ) : null}
                  {!hideDeleteBtn ? (
                    <OverlayTrigger
                      placement="top"
                      overlay={!enableDeleteBtn && deleteDisabled ? deleteDisabled() : deleteTooltip()}
                    >
                      <Button disabled={!enableDeleteBtn} onClick={removeWithConfirmation} className="admin-icons">
                        <span className="assembl-icon-delete grey" />
                      </Button>
                    </OverlayTrigger>
                  ) : null}
                </div>
              </div>
              <div className="clear" />
              <div
                className={classNames({
                  'form-tree-item': isTree,
                  'panel-body': usePanels,
                  hidden: usePanels && idx !== this.state.activePanel
                })}
              >
                {renderFields({ name: fieldname, idx: idx, fieldIndex: fieldIndex })}
                {isTree && subFieldName ? (
                  <FieldArrayWithActions
                    isTree
                    name={`${fieldname}.${subFieldName}`}
                    subFieldName={subFieldName}
                    renderFields={renderFields}
                    titleMsgId={titleMsgId}
                    renderTitleMsg={renderTitleMsg}
                    tooltips={{
                      addTooltip: addTooltip,
                      deleteTooltip: deleteTooltip,
                      deleteDisabled: deleteDisabled
                    }}
                    withSeparators={withSeparators}
                    level={level + 1}
                    maxLevel={maxLevel}
                    parentId={fieldValue.id}
                    parents={indexes}
                    confirmDeletion={confirmDeletion}
                    confirmDeletionMessages={{
                      confirmDeletionTitle: confirmDeletionTitle,
                      confirmDeletionBody: confirmDeletionBody
                    }}
                    onAdd={onAdd}
                    onRemove={onRemove}
                    onUp={onUp}
                    onDown={onDown}
                  />
                ) : null}
              </div>
              {displaySeparator ? <div className="separator" /> : null}
            </div>
          );
        })}
        {!addBtnTop ? addBtn : null}
      </div>
    );
  }
}

const FieldArrayWithActions = (props: FieldArrayProps) => {
  const { name, ...fieldsProps } = props;
  // $FlowFixMe
  return <FieldArray name={name}>{({ fields }) => <Fields fields={fields} {...fieldsProps} />}</FieldArray>;
};

FieldArrayWithActions.defaultProps = {
  usePanels: false,
  renderTitleMsg: ({ titleMsgId, idx }) => <Translate value={titleMsgId} count={idx + 1} />,
  withSeparators: true,
  subFieldName: '',
  isTree: false,
  level: 0,
  minItems: -1,
  maxLevel: MAX_TREE_FORM_LEVEL,
  parents: [],
  parentId: '',
  confirmDeletion: false,
  confirmDeletionMessages: {
    confirmDeletionTitle: () => <Translate value="deleteConfirmation.confirmDeletionTitle" />,
    confirmDeletionBody: () => <Translate value="deleteConfirmation.confirmDeletionBody" />
  }
};

export default FieldArrayWithActions;