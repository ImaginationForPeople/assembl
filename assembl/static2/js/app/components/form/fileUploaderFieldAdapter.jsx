// @flow
import * as React from 'react';
import { type FieldRenderProps } from 'react-final-form';
import { ControlLabel, FormGroup, Tooltip } from 'react-bootstrap';

import FileUploader from '../common/fileUploader';
import Error from './error';
import { getValidationState } from './utils';

type Props = {
  deleteTooltip: ?React.Element<Tooltip>,
  label: string
} & FieldRenderProps;

const FileUploaderFieldAdapter = ({ deleteTooltip, input, label, meta: { error, touched } }: Props) => {
  const onChange = (value) => {
    input.onChange({
      externalUrl: value,
      imgTitle: value.name,
      mimeType: value.type
    });
  };
  return (
    <FormGroup controlId={input.name} validationState={getValidationState(error, touched)}>
      <ControlLabel>{label}</ControlLabel>
      <FileUploader
        deleteFileTooltip={deleteTooltip}
        key={input.value ? 'notEmpty' : 'empty'}
        fileOrUrl={input.value && input.value.externalUrl}
        imgTitle={input.value && input.value.title}
        handleChange={onChange}
        mimeType={input.value && input.value.mimeType}
        name={input.name}
        isAdminUploader
        onDeleteClick={() => input.onChange('')}
      />
      <Error name={input.name} />
    </FormGroup>
  );
};

export default FileUploaderFieldAdapter;