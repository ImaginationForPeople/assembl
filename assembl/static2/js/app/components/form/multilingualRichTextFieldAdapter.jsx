// @flow
import React from 'react';
import { type FieldRenderProps } from 'react-final-form';
import { ControlLabel, FormGroup } from 'react-bootstrap';

import RichTextEditor from '../common/richTextEditor';
import Error from './error';
import { getValidationState } from './utils';

type multilingualValue = { [string]: string };

type Props = {
  editLocale: string,
  input: {
    name: string,
    onBlur: (?SyntheticFocusEvent<*>) => void,
    onChange: (SyntheticInputEvent<*> | any) => void,
    onFocus: (?SyntheticFocusEvent<*>) => void,
    value: multilingualValue
  },
  label: string
} & FieldRenderProps;

const RichTextFieldAdapter = ({
  editLocale,
  input: { name, onChange, value, ...otherListeners },
  label,
  meta: { error, touched },
  ...rest
}: Props) => {
  const valueInLocale = value[editLocale] || null;
  const key = valueInLocale ? 'notEmpty' : 'empty';
  return (
    <FormGroup controlId={name} validationState={getValidationState(error, touched)}>
      {valueInLocale ? <ControlLabel>{label}</ControlLabel> : null}
      <RichTextEditor
        {...otherListeners}
        {...rest}
        key={key}
        rawContentState={valueInLocale}
        placeholder={label}
        toolbarPosition="bottom"
        updateContentState={cs => onChange({ ...value, [editLocale]: cs })}
        withAttachmentButton={false}
      />
      <Error name={name} />
    </FormGroup>
  );
};

export default RichTextFieldAdapter;