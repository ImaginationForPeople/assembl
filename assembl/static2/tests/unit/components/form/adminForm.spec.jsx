import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import { DumbAdminForm } from '../../../../js/app/components/form/adminForm';

describe('DumbAdminForm component', () => {
  const handleSubmitSpy = jest.fn();

  it('should render an admin form', () => {
    const setRouteLeaveHookSpy = jest.fn();
    const props = {
      handleSubmit: handleSubmitSpy,
      pristine: false,
      routes: [],
      router: {
        setRouteLeaveHook: setRouteLeaveHookSpy
      },
      submitting: false
    };

    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<DumbAdminForm {...props}>nothing</DumbAdminForm>);
    const result = shallowRenderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});