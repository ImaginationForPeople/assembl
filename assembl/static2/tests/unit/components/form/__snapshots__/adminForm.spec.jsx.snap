// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DumbAdminForm component should render an admin form 1`] = `
<form
  onSubmit={[Function]}
>
  <SaveButtonInPortal
    disabled={false}
    saveAction={[Function]}
  />
  nothing
</form>
`;
